### PR TITLE
Add Multiple ports to ServiceImport

### PR DESCRIPTION
**Pull request by vthapar**
If service has multiple ports, add them all to ServiceImport,
not just first one.

Signed-off-by: Vishal Thapar <5137689+vthapar@users.noreply.github.com>

### CODE REVIEW

**Review by skitt — CHANGES_REQUESTED** [review]
Any chance this could have an associated unit test?

**Comment by tpantelis** [comment]
> Any chance this could have an associated unit test?

yeah https://github.com/submariner-io/lighthouse/pull/451 adds test for service ports. We should merge that first and add another test case here.

File: pkg/agent/controller/agent.go [hunk 1]
```diff
@@ -425,11 +425,14 @@ func (a *Controller) newServiceImport(svcExport *mcsv1a1.ServiceExport) *mcsv1a1
...
-		mcsPort.Protocol = service.Spec.Ports[0].Protocol
-		mcsPort.Port = service.Spec.Ports[0].Port
+	mcsPorts := []mcsv1a1.ServicePort{}
+
+	for _, port := range service.Spec.Ports {
+		mcsPort := mcsv1a1.ServicePort{}
+		mcsPort.Name = port.Name
+		mcsPort.Protocol = port.Protocol
+		mcsPort.Port = port.Port
+		mcsPorts = append(mcsPorts, mcsPort)
```
Comment (tpantelis) [inline pkg/agent/controller/agent.go]:
```suggestion
		mcsPorts = append(mcsPorts, mcsv1a1.ServicePort {
		    Name:     port.Name,
		    Protocol: port.Protocol,
		    Port:        port.Port,
		})
```

**Review by tpantelis — CHANGES_REQUESTED** [review]

File: pkg/agent/controller/agent.go [hunk 2]
```diff
@@ -425,11 +425,14 @@ func (a *Controller) newServiceImport(svcExport *mcsv1a1.ServiceExport) *mcsv1a1
 
 func (a *Controller) getIPsAndPortsForService(service *corev1.Service, siType mcsv1a1.ServiceImportType) (
 	[]string, []mcsv1a1.ServicePort, error) {
-	mcsPort := mcsv1a1.ServicePort{}
-	if len(service.Spec.Ports) > 0 {
-		mcsPort.Name = service.Spec.Ports[0].Name
-		mcsPort.Protocol = service.Spec.Ports[0].Protocol
-		mcsPort.Port = service.Spec.Ports[0].Port
+	mcsPorts := []mcsv1a1.ServicePort{}
```
Comment (tpantelis) [inline pkg/agent/controller/agent.go]: So we return nil if there's no ports.

```suggestion
	var mcsPorts []mcsv1a1.ServicePort
```

File: pkg/agent/controller/agent.go [hunk 3]
```diff
@@ -451,7 +454,7 @@ func (a *Controller) getIPsAndPortsForService(service *corev1.Service, siType mc
 		return make([]string, 0), make([]mcsv1a1.ServicePort, 0), nil
```
Comment (tpantelis) [inline pkg/agent/controller/agent.go]:
```suggestion
		return make([]string, 0), nil, nil
```

**Review by tpantelis — CHANGES_REQUESTED** [review]

File: pkg/agent/controller/agent.go (hunk 2 above)
Comment (vthapar) [reply pkg/agent/controller/agent.go:428]:
```
type ServiceImportSpec struct {
	// +listType=atomic
	Ports []ServicePort `json:"ports"`
```
`ServicePort` is not `omitempty` but individual fields are. So we want it to be non nil.

**Comment by tpantelis** [comment]
> > Any chance this could have an associated unit test?
> 
> yeah #451 adds test for service ports. We should merge that first and add another test case here.

@vthapar since #451 was merged, can you change the service port test case to have 2 ports to cover these changes?

**Review by aswinsuryan — APPROVED** [review]

File: pkg/agent/controller/agent_test.go [hunk 4]
```diff
@@ -182,6 +182,11 @@ var _ = Describe("ServiceImport syncing", func() {
 					Protocol: corev1.ProtocolTCP,
 					Port:     123,
 				},
+				{
```
Comment (skitt) [inline pkg/agent/controller/agent_test.go:185]: AFAICT, the test doesn’t actually verify that the ports have been copied, does it?

**Review by skitt — CHANGES_REQUESTED** [review]

**Comment by vthapar** [comment]
> Any chance this could have an associated unit test?

Done

File: pkg/agent/controller/agent_test.go (hunk 4 above)
Comment (vthapar) [reply pkg/agent/controller/agent_test.go:185]: It does. Tom already added that check as part of UT
```
	Expect(serviceImport.Spec.Ports).To(HaveLen(len(service.Spec.Ports)))

	for i := range service.Spec.Ports {
		Expect(serviceImport.Spec.Ports[i].Name).To(Equal(service.Spec.Ports[i].Name))
		Expect(serviceImport.Spec.Ports[i].Protocol).To(Equal(service.Spec.Ports[i].Protocol))
		Expect(serviceImport.Spec.Ports[i].Port).To(Equal(service.Spec.Ports[i].Port))
	}
```

**Review by skitt — APPROVED** [review]

**Review by tpantelis — APPROVED** [review]

**Pull request merged**